### ISSUE TITLE
posix_fallocate detection (TORRENT_HAS_FALLOCATE)

### ISSUE THREAD

**Issue by libTorrentUser** [arvidn/libtorrent#4801]
`TORRENT_HAS_FALLOCATE` is being "incorrectly" set to 0 on some systems where `posix_fallocate` is available. 

I confirmed the problem is happening on Alpine Linux, but I believe it will also happen on any system that uses `musl libc` because, although I believe it is supposed to conform to POSIX 2008 (`_POSIX_C_SOURCE=200809L`), it seems that the symbol is not automatically defined. Nor is `_XOPEN_SOURCE`. Therefore

[config.hpp line 179](https://github.com/arvidn/libtorrent/blob/55d9ee3451a480facb58935515236c79b1f9e097/include/libtorrent/config.hpp#L179)

    // posix_fallocate() is available under this condition
    #if _XOPEN_SOURCE >= 600 || _POSIX_C_SOURCE >= 200112L
    #define TORRENT_HAS_FALLOCATE 1
    #else
    #define TORRENT_HAS_FALLOCATE 0
    #endif

will set `TORRENT_HAS_FALLOCATE` to 0

This issue happens on the 1_2 branch. It was tested on 1.2.7 but I believe it also happens on 2_0 because the relevant parts of `config.hpp` remains the same.

Although I believe `musl libc` is the one to blame here (I think the library should be the one defining those two symbols in case they are not set by the user), since it does not, I think it would be safer to just surround the above with 

    #ifndef TORRENT_HAS_FALLOCATE
    ...
    #endif  // #ifndef TORRENT_HAS_FALLOCATE

this way, users building on platforms that do not define `_XOPEN_SOURCE` or `_POSIX_C_SOURCE` can simply define `TORRENT_HAS_FALLOCATE` in order to make `posix_fallocate` available. 

It is true that the user could simply define one of those two symbols to some appropriate value and we would not need to change a thing inside `config.hpp`, but I believe `TORRENT_HAS_FALLOCATE` has a very well defined and narrow scope, while defining any of those other symbols to wrong values may end up breaking way more stuff.

And last, but not least, maybe we could return an error or an alert or something to an user that sets `lt::storage_mode_t::storage_mode_allocate` in a system that cannot  (or was not configured to) use any of the currently available methods libtorrent uses to pre-allocate files. I believe this is particularly important on Linux, where `ls` doesn't usually report the real file size, unlike `du -h`.

**Comment by arvidn** [comment]
Anyway, apart from that; I agree that it would be good to be able to force `TORRENT_HAS_FALLOCATE` at build time. I would also like to fix the configure header to work with musl as well.

Do you have any ideas of how to detect it?
Has every version of musl supported this function, or would I have to detect the version as well?

alerting on failing to allocate is a bit tricky. On a system (or filesystem) where you know it's not supported, such alerts would just be noise. I think there may be some complications to extend the alert infrastructure down to the disk I/O code.

**Comment by arvidn** [comment]
could you please test this patch? https://github.com/arvidn/libtorrent/pull/4804

**Comment by libTorrentUser** [comment]
Sorry, but I have no idea how to detect it. And I am no musl expert. Far from it. I just noticed libtorrent was not pre-allocating properly on Alpine Linux and I knew Alpine used musl. Reading musl docs it seems that they expect the user to define the symbols, which would lead us back to your config.hpp being correct in the first place :) 

In fact, I am not even sure how exactly those symbols are defined on other platforms. I know for sure that GCC is defining them somehow when I build using the same makefile on Arch Linux. Some stackoverflow posts seem to indicate that GCC defines them when you use options like `-std=c99`, which I don't. [A quick test](https://godbolt.org/z/vCGg4n) showed that _POSIX_C_SOURCE seems to be defined only after the inclusion of some headers. 

    #include <cstdio>

    int test(void) 
    {
        return _POSIX_C_SOURCE;
    }

As that test shows, if you comment the `#include`, compilation fails. 

Anyway, although I cannot test it right now (I will try to do it tomorrow), I believe your patch would work. But I liked the way you were previously performing the check better. Seemed cleaner to me. The only "problem" it had was that it would change the definition of `TORRENT_HAS_FALLOCATE` in case the user had tried to force it, and that was why I thought it would be a good idea to only perform that check in case `TORRENT_HAS_FALLOCATE` was not already defined. That is why I suggested simply surrounding the original code with the `ifndef TORRENT_HAS_FALLOCATE ...`


As for the alerts, I wasn't really thinking about alerting in case the allocation failed, but instead do it when a torrent requesting pre-allocation is added to the session. I really don't know how viable it would be, but I was thinking we could simply check for the existence of any of those functions you use when resizing the file [here](https://github.com/arvidn/libtorrent/blob/55d9ee3451a480facb58935515236c79b1f9e097/src/file.cpp#L1167). So, if the torrent is requesting to be pre-allocated, but we are on Linux, we know it will only work if `F_PREALLOCATE` or `F_ALLOCSP64` or `TORRENT_HAS_FALLOCATE` is defined. Right?

**Comment by arvidn** [comment]
From what I can tell, musl deliberately don’t have feature test macros to detect it

**Comment by libTorrentUser** [comment]
So it seems.

By the way, is there a way to pass TORRENT_HAS FALLOCATE=1 to cmake? Although your patch seems to be working (I applied it to my project and tested it), I wanted to test it on an unblemished copy of libtorrent source, but I am very not familiar with cmake. I can build the library using the instructions on the project page, but I could not figure out a way to pass TORRENT_HAS_FALLOCATE to the compiler without modifying CMakeLists.txt directly, which I am pretty sure is not the way to go.

**Comment by arvidn** [comment]
> By the way, is there a way to pass TORRENT_HAS FALLOCATE=1 to cmake?

Not supported by the CMakeList.txt, but I would imagine you can inject flags somehow into cmake. `CMAKE_CXX_FLAGS` maybe?

**Comment by FranciscoPombal** [comment]
@libTorrentUser 

> without modifying CMakeLists.txt directly, which I am pretty sure is not the way to go.

In this case it would probably be the better, more surefire way - a simple `target_compile_definitions(torrent-rasterbar PUBLIC TORRENT_HAS_FALLOCATE)` anywhere at top-level after the `add_library(torrent-rasterbar ...` call should do the trick.

**Comment by libTorrentUser** [comment]
Thanks guys.

So,
-  I downloaded a clean copy of the source (1_2_7) 
- applied the patch. 
- made sure the library was built with TORRENT_HAS_FALLOCATE=1 (at least build.ninja displayed it among the other build definitions).
- fired client_test with `./client_test -s /tmp/download -a allocate /tmp/some.torrent`
- waited until one piece had finished downloading
- ran `du -h /tmp/download`

and the total size on disk of the 500MB torrent was only 7MB.

Did I do anything wrong?

**Comment by arvidn** [comment]
> and the total size on disk of the 500MB torrent was only 7MB.

The file size was 7 MB but it used 500 MB of disk space?

**Comment by libTorrentUser** [comment]
The torrent has a single 500MB file. 
After a couple of pieces were downloaded, it used 7MB of disk space. 
The expected was 500MB.

Even though `du -h` showed just 7MB, the `ls` command showed 500MB, indicating that a sparse file had been created.

**Comment by arvidn** [comment]
I see, I misread. I could run the same test on ubuntu. `strace` might help too

**Comment by arvidn** [comment]
as far as I can tell, `posix_fallocate()` works on ubuntu 20.04 on ext4. Immediately after stepping over the `posix_fallocate()` call, I see this:

```
$ du -h ubuntu-19.10-desktop-amd64.iso
2,3G    ubuntu-19.10-desktop-amd64.iso

2463842304
```

However, if `posix_fallocate()` fails with `EINVAL` or `ENOTSUPP`, the error is ignored and not reported by libtorrent (because those errors indicate the underlying filesystem doesn't support it)

**Comment by libTorrentUser** [comment]
Humm...
I reverted the code back to what it was before your patch and, instead of `TORRENT_HAS_FALLOCATE=1` I used `_POSIX_C_SOURCE=200112`

Then I rebuilt and repeated the test (using `client_test`) and it worked! 

Could it be that `posix_fallocate()` on Alpine only works when `_POSIX_C_SOURCE` is set to the minimal required value? I'll investigate it further and come back to post my findings.

**Comment by libTorrentUser** [comment]
Alright. I started it all from scratch. 
- Started a brand new Alpine Linux VM
- cloned the current source
- added `TORRENT_HAS_FALLOCATE=1` to CMakeLists.txt
- built
- performed the test again

and now it worked as expected. The file was fully allocated. I really don't know what might have happened before. I must have done something stupid.

**But I did notice one very important thing**
although `_XOPEN_SOURCE` and `_POSIX_C_SOURCE` are not added automatically, they may be defined deep inside some system header. I had already noticed that with `_POSIX_C_SOURCE`, which may be automatically defined by simply including `<cstdio>` in some platforms, as shown [here](https://github.com/arvidn/libtorrent/issues/4801#issuecomment-650814790).

But that is the problem. Two problems, actually.

First:
since it appears there is no standard `<features.h>` header or something like that, merely changing the include order might break a build in case the symbol is defined only after `"libtorrent/config.hpp"` is seen. 

What is worse is that the behavior changes on different platforms. For instance, the old code was able to use posix_fallocate() just fine on Arch Linux, while the call was disabled on Alpine Linux, even though posix_fallocate() is available on Alpine too. 

**Things get so messy that `_XOPEN_SOURCE` is defined by some header on Alpine too (set to 600), but is not seen until it is too late!** Just modify `simple_client.cpp` and tell it to print `_XOPEN_SOURCE` and it will print 600 on Alpine, which should have made the old `libtorrent/config.hpp` set `TORRENT_HAS_FALLOCATE` to 1, but it didn't because the symbol was not set when that header was first seen.


Second:
`_XOPEN_SOURCE` and `_POSIX_C_SOURCE` [( and the seemingly thousands of similar symbols](https://man7.org/linux/man-pages/man7/feature_test_macros.7.html)) should only be defined to a value in case the platform fully supports everything that value entails. So, in the unlikely case libtorrent is built on a platform that supports `posix_fallocate()` but does not support some other feature, `_XOPEN_SOURCE` or `_POSIX_C_SOURCE` might be set to a lower value and `TORRENT_HAS_FALLOCATE` will be set to 0. 


**Based on all that, my conclusion is**
revert the `"libtorrent/config.hpp` code back to what it was and just surround it with `ifndef TORRENT_HAS_FALLOCATE`, like this

    #ifndef TORRENT_HAS_FALLOCATE
    #if _XOPEN_SOURCE >= 600 || _POSIX_C_SOURCE >= 200112L
    #define TORRENT_HAS_FALLOCATE 1
    #else
    #define TORRENT_HAS_FALLOCATE 0
    #endif
    #endif // TORRENT_HAS_FALLOCATE

Your old code looks cleaner and, as long as it checks if the user has already set TORRENT_HAS_FALLOCATE, it will do the right thing in case `posix_fallocate()` is available on systems where `_XOPEN_SOURCE` and `_POSIX_C_SOURCE` are set to smaller values. Yes, it mail fail to compile in case `posix_fallocate()` is not available, but that will only happen if the user explicitly tried to tell the compiler it is available. libtorrent's automatic detection will do the right thing if left alone. I think that is better than your new code, which may incorrectly **silently** disable the feature.

As a bonus, C++17 can use `__has_include` to check if a header file exists. And it seems that both `glibc` and `musl-libc` have the `<features.h>` header, which is where the `_XOPEN_SOURCE` and `_POSIX_C_SOURCE` symbols are defined. So maybe you could also include that on top of `"libtorrent/config.hpp"`. It would certainly make it even more robust.

**Comment by arvidn** [comment]
I kind of prefer the new approach in libtorrent; assume the platform is posix, unless it can be detected not to be. That seems to also be the spirit of musl, they want you to assume posix is there and report bugs against them when it isn't.

I think it puts the burden of fixing build issues in the right place too, fringe platforms that don't support posix. They should be less common, and so it seems safer to default to the common case.

**Comment by libTorrentUser** [comment]
I understand your reasoning, but I think it would only make sense if the compilers automatically defined those feature symbols. The problem is that they don't and, currently, libtorrent relies on sheer luck to check posix conformance. And to make things worse, **it will silently disable `TORRENT_HAS_FALLOCATE`** and the user will never know because the library won't give him an error when he requests an allocation mode that was not compiled into the binary. 

It is your call, but I think you are making a mistake here. Full posix conformance is not needed, only the presence of a function called `posix_fallocate()`.  Although I like the old code better (I think it looks better), the new code still suffers the same problem - the code is likely to set `TORRENT_HAS_FALLOCATE=0` even on platforms will full posix conformance.

Therefore, whether you decided to revert it back to the old way or not, I urge you to at least consider surrounding that code block with `#ifndef TORRENT_HAS_FALLOCATE`. That way an user can still force the library to use `posix_fallocate()` without modifying the source by simply defining that symbol.

**Comment by arvidn** [comment]
> I understand your reasoning, but I think it would only make sense if the compilers automatically defined those feature symbols. The problem is that they don't and, currently, libtorrent relies on sheer luck to check posix conformance. And to make things worse, **it will silently disable `TORRENT_HAS_FALLOCATE`** and the user will never know because the library won't give him an error when he requests an allocation mode that was not compiled into the binary.

Since I landed my patch, libtorrent will now assume that `posix_fallocate()` is available by default. If no macros are defined (as is the case with musl for instance), `posix_fallocate()` will be used. Maybe I'm misunderstanding what you mean. As far as I can tell, the problem would be if the `_XOPEN_SOURCE` macro was *incorrectly* defined. Then the use of `posix_fallocate()` would be disabled.

Are thee other cases where it would be silently disabled?

> It is your call, but I think you are making a mistake here. Full posix conformance is not needed, only the presence of a function called `posix_fallocate()`. Although I like the old code better (I think it looks better), the new code still suffers the same problem - the code is likely to set `TORRENT_HAS_FALLOCATE=0` even on platforms will full posix conformance.

How?

Just to be clear, I think we're talking about [this](https://github.com/arvidn/libtorrent/blob/RC_1_2/include/libtorrent/config.hpp#L180) logic for android.

```
// posix_fallocate() is not available in glibc under these condition
#if defined _XOPEN_SOURCE && _XOPEN_SOURCE < 600
#define TORRENT_HAS_FALLOCATE 0
#elif defined _POSIX_C_SOURCE && _POSIX_C_SOURCE < 200112L
#define TORRENT_HAS_FALLOCATE 0
#endif
```

combined with the default fallback [here](https://github.com/arvidn/libtorrent/blob/RC_1_2/include/libtorrent/config.hpp#L373):

```
#ifndef TORRENT_HAS_FALLOCATE
#define TORRENT_HAS_FALLOCATE 1
#endif
```

There are a few other platforms that are known to now have `posix_fallocate()`, like OS/2, BeOS, windows and MacOS. So those also define it to 0. But I don't think that's what we're talking about, right?

> Therefore, whether you decided to revert it back to the old way or not, I urge you to at least consider surrounding that code block with `#ifndef TORRENT_HAS_FALLOCATE`. That way an user can still force the library to use `posix_fallocate()` without modifying the source by simply defining that symbol.

Yes, it would be nice to have a way to force this from the command line. I would be a bit tempted to have a different macro that's settable by the user though. If I recall the convention I've used so far would make that `TORRENT_USE_FALLOCATE`. So, the `HAS` macro is what the detection logic found, and `USE` would be what's actually used.

**Comment by libTorrentUser** [comment]
Ha! I stand corrected. Once again :)

I could swear the default fallback was doing `TORRENT_HAS_FALLOCATE=0`, and that is why I was being so annoying, after all, your patch only touched that other code block. I don't know how I missed that, nor why my local copy was setting it to zero. I may have done something stupid while testing. Your changes makes much more sense now.

But... although that solves the problem on my side, aren't we reversing it now? What if the platform does not have `posix_fallocate()`?  

Including the `<features.h>` header (in case it is available) somewhere inside `config.hpp` would ensure `_POSIX_C_SOURCE` (and its family of symbols) are visible when the pre-processor reaches that block. But I guess the problem would still remain in case this obscure hypothetical platform does not have the function and does not define those symbols. 

In the end, I think the best way would be to give the user a choice to force it from the command line. And I liked your idea of using `HAS` and `USE` . If you were already doing it somewhere else, I guess this is another good candidate.


So... although you've pretty much solved my problem, I guess we can still improve the code by

- including the `<features.h>` header inside `config.hpp`, in case it exists.
- going with your idea and allowing the user to pass `TORRENT_USE_FALLOCATE` in the command line and only modify it inside `config.hpp` in case it was not previously defined.

What do you think?

**Comment by arvidn** [comment]
> What if the platform does not have posix_fallocate()?

I think such platforms are in minority, and it makes sense for the user to have to tinker a bit to get it to build.

> including the <features.h> header inside config.hpp, in case it exists.

Yes, the `master` branch is now requiring C++17, so it could be done there.

> going with your idea and allowing the user to pass TORRENT_USE_FALLOCATE in the command line and only modify it inside config.hpp in case it was not previously defined.

I agree that this would be an improvement, it's not very high on my list right now though.

**Comment by libTorrentUser** [comment]
Great! As usual, I thank you for your your time and patience. 

Would you like me to close this issue now or you'd rather it be left open until either you or I have enough time to submit a patch that deals with including the <features.h> header and the TORRENT_USE_FALLOCATE?

**Comment by arvidn** [comment]
I think since the core issue of this ticket has been resolved, adding the ability to force using fallocate on or off should have its own ticket.
